### PR TITLE
logical And Or search functionality added

### PR DESCRIPTION
**Pull request by chenpionaar1998**

### CODE REVIEW

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 1]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
     */
-  public Set<Message> searchWord (String word) {
-    if (wordsHash.containsKey(word)) {
-      return wordsHash.get(word);
+  public List<Message> search (String searchTarget) {
+    if (searchTarget.contains("||")) {
+      return searchUnion(searchTarget);
+    }else if (searchTarget.contains("&&")) {
+      return searchIntersection(searchTarget);
+    }else if (wordsHash.containsKey(searchTarget)){
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: Does it make sense to trim to first word here if the searchTarget has a whitespace?  The hash only contains single words right?

Comment (turtlefarmer) [reply src/main/java/codeu/model/store/basic/IndexStore.java]: Assuming if we search for "hello world" we'd rather at least return results for hello than nothing.
Alternatively, "hello world" could be parsed to automatically be "hello && world""

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 2]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
     return null;
   }
 
+  /**
+    * Pre: the given string contains "||"
+    * returns the union of the elements in the given searchTarget
+    */
+  public List<Message> searchUnion (String searchTarget) {
+    // remove bad format || in searchtarget
+    if (searchTarget.indexOf("|") == 0) {
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: Instead of all of this logic.  Is it functionally equivalent if you just ignore results in words that is empty string?  That would have the added benefit of handling cases such as "a || || b" and still returning a valid result.

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 3]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
+    }
+    if (searchTarget.indexOf("||") == (searchTarget.length()-2)) {
+      searchTarget = searchTarget.substring(0,searchTarget.length()-2);
+    }
+    String[] words = searchTarget.split("\\|\\|");
+
+    if (words.length == 0) {
+      return null;
+    }else {
+      Set<Message> hashSetResult = new HashSet<Message>(search(words[0]));
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: Does anything bad here happen if search returns null?

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 4]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
+      for (String word : words){
+        if (search(word) != null) {
+          Set<Message> hashSetTemp = new HashSet<Message>(search(word));
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: There's a redundant call of search(word) here.  It would be bad if search(word) is expensive.

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 5]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
+      searchTarget = searchTarget.substring(2);
+    }
+    if (searchTarget.indexOf("||") == (searchTarget.length()-2)) {
+      searchTarget = searchTarget.substring(0,searchTarget.length()-2);
+    }
+    String[] words = searchTarget.split("\\|\\|");
+
+    if (words.length == 0) {
+      return null;
+    }else {
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: You don't need the else here if the if block always returns.  Code to remove levels of indentation whenever possible for readability.

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 6]
```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
...
+      unionResult = sortMessageList(unionResult);
+      return unionResult;
+    }
+  }
+
+  /**
+    * Pre: the given string contains "&&"
+    * returns the intersection of the elements in the given searchTarget
+    */
+  public List<Message> searchIntersection (String searchTarget) {
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java:118]: Pretty much identical feedback to searchUnion

File: src/main/java/codeu/model/store/basic/IndexStore.java (hunk 3 above)
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: Consider trimming items in words here.  Then you code would also be able to support "hello || world"

File: src/test/java/codeu/model/store/basic/IndexStoreTest.java [hunk 7]
```diff
@@ -70,4 +100,35 @@ public void testSplitAndHashMessage () {
...
+    Assert.assertEquals(expectedList, indexStore.search("test&&one&&message"));
+    Assert.assertEquals(expectedList, indexStore.search("test&&one&&"));
+    Assert.assertEquals(expectedList, indexStore.search("&&one&&test"));
+
+  }
+
+  @Test
+  public void testSearchUnion () {
+    List<Message> expectedList = new ArrayList<>();
+    expectedList.add(MESSAGE_TWO);
```
Comment (turtlefarmer) [inline src/test/java/codeu/model/store/basic/IndexStoreTest.java:135]: Add union and intersection tests for case when the expectedList is empty or null

**Review by turtlefarmer — APPROVED** [review]

File: src/main/java/codeu/model/store/basic/IndexStore.java (hunk 3 above)
Comment (chenpionaar1998) [reply src/main/java/codeu/model/store/basic/IndexStore.java]: right i think it gives a null pointer exception if it's null. I think a better solution would be to create an empty one first and add things to it in the for loop under, this prevents null results in initializing and also takes care of the redundant call mentioned under

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 8]
```diff
@@ -66,13 +71,88 @@ public void addMapping (String word, Message message) {
...
     * Search function for a single word, returns a Set of messages from the Hashtable if the word is in it, otherwise return null
     */
-  public Set<Message> searchWord (String word) {
-    if (wordsHash.containsKey(word)) {
-      return wordsHash.get(word);
+  public List<Message> search (String searchTarget) {
+    if (searchTarget.contains("||")) {
+      return searchUnion(searchTarget);
+    }else if (searchTarget.contains("&&")) {
+      searchTarget = searchTarget.replaceAll(" ","");
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java]: I think this woudl be nice if it also replaced whitespace with &&s.  Currently this turns "hello && world foo" into "hello && worldfoo"  I think implied conversion into "hello && world && foo" makes more sense.

File: src/main/java/codeu/model/store/basic/IndexStore.java [hunk 9]
```diff
@@ -66,13 +71,88 @@ public void addMapping (String word, Message message) {
...
+  public List<Message> searchIntersection (String searchTarget) {
+    if (searchTarget.indexOf("&&") == 0) {
+      searchTarget = searchTarget.substring(2);
+    }
+    String[] words = searchTarget.split("&&");
+    if (words.length == 0 || search(words[0]) == null) {
+      return null;
+    }
+    List<Message> intersectionResult = new ArrayList<Message>(search(words[0]));
+    for (int i = 1 ; i < words.length ; i++) {
```
Comment (turtlefarmer) [inline src/main/java/codeu/model/store/basic/IndexStore.java:127]: Any reason why this isn't also a for iterator loop like in the || case above?

**Review by turtlefarmer — APPROVED** [review]

File: src/main/java/codeu/model/store/basic/IndexStore.java (hunk 9 above)
Comment (chenpionaar1998) [reply src/main/java/codeu/model/store/basic/IndexStore.java:127]: yeah, I can't initialize the result to be empty since the retainAll function compares the two list and removes items from the caller. So in order to not have a redundant call for search like mentioned, I started the loop from i = 1
